### PR TITLE
Trim whitespace from address fields before database save

### PR DESCRIPTION
**Pull request by ProxiBlue**
## Summary

- Add `trimAddressFields()` to `AbstractAddress::beforeSave()` to trim leading/trailing whitespace (including Unicode) from all string address fields before persistence
- Covers both checkout (`Quote\Address`) and customer address (`Customer\Address`) save paths since both extend `AbstractAddress`
- Uses `preg_replace` with `/u` flag to also strip Unicode whitespace (non-breaking spaces, zero-width spaces)
- Includes 9 unit tests covering all trimmed fields, street handling, Unicode whitespace, null safety, and edge cases

Fixes https://github.com/mage-os/mageos-magento2/issues/214

## Fields trimmed

`firstname`, `lastname`, `middlename`, `prefix`, `suffix`, `company`, `city`, `region`, `telephone`, `fax`, `postcode`, `vat_id`, `email`, and each line of the `street` field.

## Test plan

- [ ] Verify checkout (guest): enter address fields with leading/trailing spaces → confirm trimmed values in `sales_order_address`
- [ ] Verify customer address add/edit: enter fields with spaces → confirm trimmed on save
- [ ] Verify existing orders/addresses are unaffected (trim only runs on new saves)
- [ ] Run unit tests: `vendor/bin/phpunit -c dev/tests/unit/phpunit.xml.dist app/code/Magento/Customer/Test/Unit/Model/Address/AbstractAddressTest.php`

### CODE REVIEW

**Comment by rhoerr** [comment]
Thanks for the PR

**Comment by rhoerr** [comment]
Can you change the PR base to `release/3.x` for Mage-OS 3.0? You might need to remerge in the process.

Thanks

File: app/code/Magento/Customer/Model/Address/AbstractAddress.php [hunk 1]
```diff
@@ -611,10 +611,76 @@ public function getConfig()
     public function beforeSave()
     {
         parent::beforeSave();
+        $this->trimAddressFields();
```
Comment (rhoerr) [inline app/code/Magento/Customer/Model/Address/AbstractAddress.php]:
```suggestion
        $this->trimAddressFields();
        parent::beforeSave();
```

It seems to me it would be better to trim before running before_save events, so that the fields are consistent between observers and what actually gets saved. Thoughts?

File: app/code/Magento/Customer/Model/Address/AbstractAddress.php [hunk 2]
```diff
@@ -611,10 +611,76 @@ public function getConfig()
...
+            'suffix',
+            'company',
+            'city',
+            'region',
+            'telephone',
+            'fax',
+            'postcode',
+            'vat_id',
+            'email',
+        ];
```
Comment (rhoerr) [inline app/code/Magento/Customer/Model/Address/AbstractAddress.php:643]: Is there a particular reason to do these specific fields, rather than (hypothetically) all text-type address attributes?

Comment (ProxiBlue) [reply app/code/Magento/Customer/Model/Address/AbstractAddress.php:643]: I did not want just change any address field, as custom module smay be affected, and I can;t predict or test if that would break their fields if I tamper with them. So figured to keep it to base known fields.

File: app/code/Magento/Customer/Model/Address/AbstractAddress.php (hunk 1 above)
Comment (ProxiBlue) [reply app/code/Magento/Customer/Model/Address/AbstractAddress.php]: That does make more sense, good catch

**Review by rhoerr — APPROVED** [review]

**Comment by ProxiBlue** [comment]
@rhoerr 

Something has just surfaced that may need this to be held back. Not sure what your schedule is on release of 3, but this bug could lead to sales lost/checkout failure, so disclosing it now in case it is not fixed in time.

I don't know if the issue is only specific to shipperHQ enhanced shipping solution, or any shipping vendor that requires address lookup api calls for results. (very possible the later)

I setup a new review process, and AI cought this and alerted me:

```
Ticket: checkout-Placed-order-trims-leading/trailing-spaces-from-address-fields — Playwright test failure
                                                                                   
  Status: Root cause identified. Action required.
                                                                                                                                                                                                                 
  ---                                                                                                                                                                                                              
  What the test does                                                                                                                                                                                               
                                                                                                                                                                                                                   
  Fills checkout address fields with leading/trailing spaces (e.g., "  TestCompany  "), places an order, and verifies the order succeeds and the stored address has those spaces stripped.                         

  ---                                                                                                                                                                                                              
  Failure observed                                                                                                                                                                                                 
                                  
  Test times out at 180 seconds. The browser is closed by the test harness before page.waitForLoadState("networkidle") resolves. The order is never placed — Magento redirects the user back to the shipping step  
  instead of showing the success page.                                                                                                                                                                             
                                                                                                                                                                                                                 
  ---                                                                                                                                                                                                              
  Root cause                      
                                                                                                                                                                                                                   
  The trim patch (patches/mage-os-module-customer-trim-address-fields.patch) runs inside AbstractAddress::beforeSave(). This fires when the shipping address is saved to the database, which happens during
  setShippingInformation (the "Next" button click at end of shipping step).
                                                                                                                                                                                                                 
  ShipperHQ encodes the carrier group into its rate codes. When the company name contains spaces (e.g., "  TestCompany  "), ShipperHQ may classify the address into a different carrier group (e.g., residential   
  vs. commercial) than it does after the spaces are trimmed. The carrier group appears in the rate code: shipperhq_1_FEDEX_GROUND vs. shipperhq_2_FEDEX_GROUND.                                                  
                                                                                                                                                                                                                   
  The failure chain:              
                                                                                                                                                                                                                   
  1. Browser calls estimate-shipping-methods with spaced address → ShipperHQ returns group-1 rates (e.g., shipperhq_1_FEDEX_GROUND). User selects one.
  2. User clicks Next → setShippingInformation fires → AbstractAddress::beforeSave() trims the company name → trim runs → "  TestCompany  " becomes "TestCompany".                                               
  3. setShippingInformation also calls collectShippingRates() internally with the now-trimmed address → ShipperHQ re-classifies → returns group-2 rates. Old group-1 rates marked deleted in memory.               
  4. A bug in Quote\Address::getShippingRateByCode() (Mage-OS core) iterates all rates including deleted ones — so the method still "finds" the group-1 rate even though it was deleted. The validation check    
  inside setShippingInformation incorrectly passes.                                                                                                                                                                
  5. Stored shipping method code in the database: shipperhq_1_FEDEX_GROUND. Stored rates in the database: group-2 only.                                                                                            
  6. At place order, Magento loads the quote fresh from the database and re-collects totals. Fresh load: only group-2 rates present. Validation: getShippingRateByCode('shipperhq_1_...') → not found → false →    
  ShippingMethodValidationRule throws "The shipping method is missing. Verify the shipping method and try again.".                                                                                                 
  7. Magento redirects to the shipping step. The test waits for networkidle on the success page URL that never loads. 180-second timeout → browser killed. 
```

failure is very specific:

```
Did this cause real sales drop-off in production?                                                                                                                                                                
                                                                                                                                                                                                                   
  Unlikely, for this reason: the mismatch only triggers when ALL of the following are simultaneously true:                                                                                                         
                                                                                                                                                                                                                   
  1. Customer submits an address field with leading/trailing spaces                                                                                                                                                
  2. Trimming those spaces causes ShipperHQ to return a different carrier group (not just different prices — a different group ID encoded in the rate code)                                                        
  3. Customer proceeds all the way to place order                                                                                                                                                                  
  
  Condition 2 is the gate. In practice, ShipperHQ's carrier group assignment is driven by warehouse origin, product weight/dimensions, and carrier availability — not the company name field. The test uses "      
  TestCompany  " which may have triggered a residential/commercial reclassification in ShipperHQ's test environment, but in production, the company field contents are unlikely to shift which warehouse group
  handles the order.                                                                                                                                                                                               

  Real customers entering spaces is also rare — browser autofill doesn't pad, manual typing doesn't pad. Copy-paste from documents can, but it's edge-case volume.                                                 
  
  Is there a real checkout failure risk going forward?                                                                                                                                                             
  
  Yes, but narrow. If a customer does submit padded values AND ShipperHQ's classification changes on trim, the order silently fails at place order with "The shipping method is missing" — no actionable error     
  message, user stranded. That is a real abandonment event for that customer. The probability in production is low but non-zero.

```

Since last week when I put this patch live, zero occurrences.
I will submit a new PR / update PR

**Comment by ProxiBlue** [comment]
point 4 is also of interest:

> 4. A bug in Quote\Address::getShippingRateByCode() (Mage-OS core) iterates all rates including deleted ones — so the method still "finds" the group-1 rate even though it was deleted. The validation check    
  inside setShippingInformation incorrectly passes.

**Comment by rhoerr** [comment]
> Something has just surfaced that may need this to be held back. Not sure what your schedule is on release of 3, but this bug could lead to sales lost/checkout failure, so disclosing it now in case it is not fixed in time.
> 
> I don't know if the issue is only specific to shipperHQ enhanced shipping solution, or any shipping vendor that requires address lookup api calls for results. (very possible the later)

Thanks for the heads up. Release target is next week, maybe Tuesday. If you can submit a fix for the issue(s) this weekend that would be swell.

**Comment by ProxiBlue** [comment]
Requesting revert of this PR. The trim-on-save approach is correct in isolation but is one piece of a larger fix chain that has not been fully validated. The interaction with `collectShippingRates()` and third-party carrier rate code stability (specifically ShipperHQ) exposes edge cases we have not fully resolved. Will re-submit a complete, tested fix. Apologies — please revert if possible.
